packages/react/src/combobox/root/utils/useFilter.ts

```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
+
+const cache = new Map<string, Intl.Collator>();
+
+function getCollator(options: Intl.CollatorOptions = {}) {
+  const optionsString = JSON.stringify(options);
+  const cachedCollator = cache.get(optionsString);
+
+  if (cachedCollator) {
+    return cachedCollator;
+  }
+
+  const collator = new Intl.Collator(undefined, options);
+  cache.set(optionsString, collator);
+
+  return collator;
+}
+
+/**
+ * Matches items against a query using `Intl.Collator` for robust string matching.
+ */
+export function useFilter(options: Intl.CollatorOptions = {}) {
+  const collator = React.useMemo(() => getCollator(options), [options]);
```

Comment:
Why `useMemo`? `getCollator` already memoizes, that's 2 levels of memoization for the same value.

Reply:
Furthermore, you could also memo the full filter object `({ contains, startsWith, endsWith })` rather than just the `collator` instance, you could avoid using any `useMemo` at all in `useFilter`.